File: app/src/main/java/org/ole/planet/myplanet/model/RealmMyCourse.java
package org.ole.planet.myplanet.model;

import android.content.SharedPreferences;
import android.text.TextUtils;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import org.ole.planet.myplanet.utilities.JsonUtils;
import org.ole.planet.myplanet.utilities.Utilities;

import java.util.ArrayList;
import java.util.List;

import io.realm.Realm;
import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.RealmResults;
import io.realm.annotations.PrimaryKey;

public class RealmMyCourse extends RealmObject {
    @PrimaryKey
    private String id;
    private RealmList<String> userId;
    private String courseId;
    private String course_rev;
    private String languageOfInstruction;
    private String courseTitle;
    private Integer memberLimit;
    private String description;
    private String method;
    private String gradeLevel;
    private String subjectLevel;
    private String createdDate;
    private Integer numberOfSteps;

    public static void insertMyCourses(String userId, JsonObject myCousesDoc, Realm mRealm) {
        Utilities.log("INSERT COURSE " + new Gson().toJson(myCousesDoc));
        String id = JsonUtils.getString("_id", myCousesDoc);
        RealmMyCourse myMyCoursesDB = mRealm.where(RealmMyCourse.class).equalTo("id", id).findFirst();
        if (myMyCoursesDB == null) {
            myMyCoursesDB = mRealm.createObject(RealmMyCourse.class, id);
        }
        myMyCoursesDB.setUserId(userId);
        myMyCoursesDB.setCourseId(JsonUtils.getString("_id", myCousesDoc));
        myMyCoursesDB.setCourse_rev(JsonUtils.getString("_rev", myCousesDoc));
        myMyCoursesDB.setLanguageOfInstruction(JsonUtils.getString("languageOfInstruction", myCousesDoc));
        myMyCoursesDB.setCourseTitle(JsonUtils.getString("courseTitle", myCousesDoc));
        myMyCoursesDB.setMemberLimit(JsonUtils.getInt("memberLimit", myCousesDoc));
        myMyCoursesDB.setDescription(JsonUtils.getString("description", myCousesDoc));
        myMyCoursesDB.setMethod(JsonUtils.getString("method", myCousesDoc));
        myMyCoursesDB.setGradeLevel(JsonUtils.getString("gradeLevel", myCousesDoc));
        myMyCoursesDB.setSubjectLevel(JsonUtils.getString("subjectLevel", myCousesDoc));
        myMyCoursesDB.setCreatedDate(JsonUtils.getString("createdDate", myCousesDoc));
        myMyCoursesDB.setnumberOfSteps(JsonUtils.getJsonArray("steps", myCousesDoc).size());
        RealmCourseStep.insertCourseSteps(myMyCoursesDB.getCourseId(), JsonUtils.getJsonArray("steps", myCousesDoc), JsonUtils.getJsonArray("steps", myCousesDoc).size(), mRealm);
    }

    public static List<RealmObject> getMyByUserId(Realm mRealm, SharedPreferences settings) {
        RealmResults<RealmMyCourse> libs = mRealm.where(RealmMyCourse.class).findAll();
        List<RealmObject> libraries = new ArrayList<>();
        for (RealmMyCourse item : libs) {
            if (item.getUserId().contains(settings.getString("userId", "--"))) {
                libraries.add(item);
            }
        }
        return libraries;
    }


    public static List<RealmMyCourse> getMyCourseByUserId(String userId, List<RealmMyCourse> libs) {
        List<RealmMyCourse> libraries = new ArrayList<>();
        for (RealmMyCourse item : libs) {
            if (item.getUserId().contains(userId)) {
                libraries.add(item);
            }
        }
        return libraries;
    }

    public static List<RealmMyCourse> getOurCourse(String userId, List<RealmMyCourse> libs) {
        List<RealmMyCourse> libraries = new ArrayList<>();
        for (RealmMyCourse item : libs) {
            if (!item.getUserId().contains(userId)) {
                libraries.add(item);
            }
        }
        return libraries;
    }


    public static boolean isMyCourse(String userId, String courseId, Realm realm) {
        return RealmMyCourse.getMyCourseByUserId(userId, realm.where(RealmMyCourse.class).equalTo("courseId", courseId).findAll()).size() > 0;
    }

    public static void insert(Realm mRealm, JsonObject doc) {
        insertMyCourses("", doc, mRealm);
    }

    public static RealmMyCourse getMyCourse(Realm mRealm, String id) {
        return mRealm.where(RealmMyCourse.class).equalTo("courseId", id).findFirst();
    }

    public static void createMyCourse(RealmMyCourse course, Realm mRealm, String id) {
        if (!mRealm.isInTransaction())
            mRealm.beginTransaction();
        course.setUserId(id);
        mRealm.commitTransaction();
    }
//
//    public static String[] getMyCourseIds(Realm mRealm, String userId) {
//        List<RealmMyCourse> list = mRealm.where(RealmMyCourse.class).equalTo("userId", userId).findAll();
//        String[] myIds = new String[list.size()];
//        for (int i = 0; i < list.size(); i++) {
//            myIds[i] = list.get(i).getCourseId();
//        }
//        return myIds;
//    }

    public static JsonArray getMyCourseIds(Realm realm, String userId) {
        List<RealmMyCourse> myCourses = getMyCourseByUserId(userId, realm.where(RealmMyCourse.class).findAll());
        JsonArray ids = new JsonArray();
        for (RealmObject lib : myCourses
        ) {
            ids.add(((RealmMyCourse) lib).getCourseId());
        }
        return ids;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public RealmList<String> getUserId() {
        return userId;
    }

    public void setUserId(String userId) {
        if (this.userId == null) {
            this.userId = new RealmList<>();
        }

        if (!this.userId.contains(userId) && !TextUtils.isEmpty(userId))
            this.userId.add(userId);
    }

    public void removeUserId(String userId) {
        this.userId.remove(userId);
    }

    public String getCourseId() {
        return courseId;
    }

    public void setCourseId(String courseId) {
        this.courseId = courseId;
    }

    public String getCourse_rev() {
        return course_rev;
    }

    public void setCourse_rev(String course_rev) {
        this.course_rev = course_rev;
    }

    public String getLanguageOfInstruction() {
        return languageOfInstruction;
    }

    public void setLanguageOfInstruction(String languageOfInstruction) {
        this.languageOfInstruction = languageOfInstruction;
    }

    public String getCourseTitle() {
        return courseTitle;
    }

    public void setCourseTitle(String courseTitle) {
        this.courseTitle = courseTitle;
    }

    public Integer getMemberLimit() {
        return memberLimit;
    }

    public void setMemberLimit(Integer memberLimit) {
        this.memberLimit = memberLimit;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getMethod() {
        return method;
    }

    public void setMethod(String method) {
        this.method = method;
    }

    public String getGradeLevel() {
        return gradeLevel;
    }

    public void setGradeLevel(String gradeLevel) {
        this.gradeLevel = gradeLevel;
    }

    public String getSubjectLevel() {
        return subjectLevel;
    }

    public void setSubjectLevel(String subjectLevel) {
        this.subjectLevel = subjectLevel;
    }

    public String getCreatedDate() {
        return createdDate;
    }

    public void setCreatedDate(String createdDate) {
        this.createdDate = createdDate;
    }

    public Integer getnumberOfSteps() {
        return numberOfSteps == null ? 0 : numberOfSteps;
    }

    public void setnumberOfSteps(Integer numberOfSteps) {
        this.numberOfSteps = numberOfSteps;
    }

    @Override
    public String toString() {
        return courseTitle;
    }
}

File: app/src/main/java/org/ole/planet/myplanet/model/RealmMyCourse.kt
package org.ole.planet.myplanet.model

import android.content.Context
import android.text.TextUtils
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import org.ole.planet.myplanet.services.SharedPrefManager
import org.ole.planet.myplanet.utils.FileUtils.getOlePath
import org.ole.planet.myplanet.utils.JsonUtils

open class RealmMyCourse {
    var id: String? = null
    var userId: MutableList<String>? = null
        private set
    var courseId: String? = null
    var courseRev: String? = null
    var languageOfInstruction: String? = null
    var courseTitle: String? = null
    var courseTitleNormal: String? = null
    var memberLimit: Int? = null
    var description: String? = null
    var method: String? = null
    var gradeLevel: String? = null
    var subjectLevel: String? = null
    var createdDate: Long = 0
    var coverFileName: String? = null
    private var numberOfSteps: Int? = null
    var courseSteps: MutableList<RealmCourseStep>? = null
    @Transient
    var isMyCourse: Boolean = false
    fun setUserId(userId: String?) {
        if (this.userId == null) {
            this.userId = mutableListOf()
        }
        if (this.userId?.contains(userId) != true && !TextUtils.isEmpty(userId)) {
            this.userId?.add(userId)
        }
    }

    fun removeUserId(userId: String?) {
        this.userId?.remove(userId)
    }

    fun getNumberOfSteps(): Int {
        return numberOfSteps ?: 0
    }

    fun setNumberOfSteps(numberOfSteps: Int?) {
        this.numberOfSteps = numberOfSteps
    }

    override fun toString(): String {
        return courseTitle ?: ""
    }

    companion object {
        private val concatenatedLinks = HashSet<String>()

        fun getCoverImageFile(context: Context, courseId: String?, fileName: String?): java.io.File? {
            if (courseId.isNullOrBlank() || fileName.isNullOrBlank()) return null
            return java.io.File(
                "${getOlePath(context)}course_attachments/$courseId/$fileName"
            )
        }

        fun addConcatenatedLink(link: String) {
            synchronized(concatenatedLinks) {
                concatenatedLinks.add(link)
            }
        }

        fun saveConcatenatedLinksToPrefs(spm: SharedPrefManager) {
            val existingJsonLinks = spm.getConcatenatedLinks()
            val existingConcatenatedLinks = if (existingJsonLinks != null) {
                JsonUtils.gson.fromJson(existingJsonLinks, Array<String>::class.java).toMutableSet()
            } else {
                mutableSetOf()
            }
            val linksToProcess: List<String>
            synchronized(concatenatedLinks) {
                linksToProcess = concatenatedLinks.toList()
            }
            val existingSet = existingConcatenatedLinks.toHashSet()
            for (link in linksToProcess) {
                existingSet.add(link)
            }
            val jsonConcatenatedLinks = JsonUtils.gson.toJson(existingSet.toList())
            spm.setConcatenatedLinks(jsonConcatenatedLinks)
        }

        fun serialize(course: RealmMyCourse, resourcesByStepId: Map<String?, List<RealmMyLibrary>>): JsonObject {
            val obj = JsonObject()
            obj.addProperty("_id", course.courseId)
            obj.addProperty("_rev", course.courseRev)
            obj.addProperty("courseTitle", course.courseTitle)
            obj.addProperty("description", course.description)
            obj.addProperty("languageOfInstruction", course.languageOfInstruction)
            obj.addProperty("gradeLevel", course.gradeLevel)
            obj.addProperty("subjectLevel", course.subjectLevel)
            obj.addProperty("createdDate", course.createdDate)
            obj.addProperty("method", course.method)
            obj.addProperty("memberLimit", course.memberLimit)
            course.coverFileName?.let { obj.addProperty("coverFileName", it) }

            val stepsArray = JsonArray()

            course.courseSteps?.forEach { step ->
                val stepObj = JsonObject()
                stepObj.addProperty("stepTitle", step.stepTitle)
                stepObj.addProperty("description", step.description)
                stepObj.addProperty("id", step.id)

                val resourcesArray = JsonArray()
                val stepResources = resourcesByStepId[step.id] ?: emptyList()

                stepResources.forEach { resource ->
                    resourcesArray.add(resource.serializeResource())
                }
                stepObj.add("resources", resourcesArray)
                stepsArray.add(stepObj)
            }
            obj.add("steps", stepsArray)
            obj.add("images", JsonArray())
            return obj
        }
    }
}
